### PR TITLE
feat(sgx): shim_address() for normal and naked functions

### PR DESCRIPTION
**Pull request by jarkkojs**
Closes: #1560
Signed-off-by: Jarkko Sakkinen <jarkko@profian.com>

### CODE REVIEW

File: internal/shim-sgx/src/lib.rs [hunk 1]
```diff
@@ -55,8 +55,17 @@ pub const BLOCK_SIZE: usize = 69632;
...
     /// Extern
     pub static ENARX_EXEC_START: u8;
     /// Extern
     pub static ENARX_EXEC_END: u8;
 }
+
+/// Get the Shim's base address. It can be taken from any symbol inside the
+/// shim, but the reason for having ENARX_SHIM_ADDRESS is that it can be easily
+/// adapted to naked functions.
+pub fn shim_address() -> usize {
```
Comment (jarkkojs) [inline internal/shim-sgx/src/lib.rs:69]: missing `#[inline]`. I'll add it.

**Comment by npmccallum** [comment]
@jarkkojs We used to have a simpler function for this. But it looks like it was removed during the gdb effort. I want @haraldh to review this just in case I missed something. But otherwise, LGTM.

**Comment by jarkkojs** [comment]
@haraldh ?

**Comment by haraldh** [comment]
> @haraldh ?

looking

File: internal/shim-sgx/src/main.rs [hunk 2]
```diff
@@ -167,10 +165,9 @@ unsafe extern "sysv64" fn relocate() {
         "push   r10",
         "push   r11",
 
-        "lea    rdi,    [rip + _DYNAMIC]", // rdi = address of _DYNAMIC section
-        "mov    rsi,    -{SIZE}         ", // rsi = enclave start address mask
-        "and    rsi,    rdi             ", // rsi = relocation address
-        "call   {DYN_RELOC}             ", // relocate the dynamic symbols
+        "lea    rdi,    [rip + _DYNAMIC]",           // rdi = address of _DYNAMIC section
+        "lea    rsi,    [rip + ENARX_SHIM_ADDRESS]", // rsi = enclave start address mask
+        "call   {DYN_RELOC}",                        // relocate the dynamic symbols
```
Comment (haraldh) [inline internal/shim-sgx/src/main.rs:170]:
```diff
diff --git a/internal/shim-sgx/src/main.rs b/internal/shim-sgx/src/main.rs
index 0530b1c0..3f205032 100644
--- a/internal/shim-sgx/src/main.rs
+++ b/internal/shim-sgx/src/main.rs
@@ -19,7 +19,9 @@ extern crate rcrt1;
 
 use core::arch::asm;
 
-use shim_sgx::{entry, handler, ATTR, BLOCK_SIZE, ENARX_EXEC_START, ENCL_SIZE_BITS, MISC};
+use shim_sgx::{
+    entry, handler, ATTR, BLOCK_SIZE, ENARX_EXEC_START, ENARX_SHIM_ADDRESS, ENCL_SIZE_BITS, MISC,
+};
 
 #[panic_handler]
 #[cfg(not(test))]
@@ -166,7 +168,7 @@ unsafe extern "sysv64" fn relocate() {
         "push   r11",
 
         "lea    rdi,    [rip + _DYNAMIC]",           // rdi = address of _DYNAMIC section
-        "lea    rsi,    [rip + ENARX_SHIM_ADDRESS]", // rsi = enclave start address mask
+        "lea    rsi,    [rip + {ENARX_SHIM_ADDRESS}]", // rsi = enclave start address mask
         "call   {DYN_RELOC}",                        // relocate the dynamic symbols
 
         "pop    r11",
@@ -182,6 +184,7 @@ unsafe extern "sysv64" fn relocate() {
         "ret",
 
         DYN_RELOC = sym rcrt1::dyn_reloc,
+        ENARX_SHIM_ADDRESS = sym ENARX_SHIM_ADDRESS,
         options(noreturn)
     )
 }
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: internal/shim-sgx/src/main.rs (hunk 2 above)
Comment (haraldh) [reply internal/shim-sgx/src/main.rs:170]: optional

Comment (jarkkojs) [reply internal/shim-sgx/src/main.rs:170]: Changes done, I'll just run `cargo test` as a sanity check.

File: internal/shim-sgx/src/main.rs [hunk 3]
```diff
@@ -167,10 +167,9 @@ unsafe extern "sysv64" fn relocate() {
         "push   r10",
         "push   r11",
 
-        "lea    rdi,    [rip + _DYNAMIC]", // rdi = address of _DYNAMIC section
-        "mov    rsi,    -{SIZE}         ", // rsi = enclave start address mask
-        "and    rsi,    rdi             ", // rsi = relocation address
-        "call   {DYN_RELOC}             ", // relocate the dynamic symbols
+        "lea    rdi,    [rip + _DYNAMIC]",             // rdi = address of _DYNAMIC section
+        "lea    rsi,    [rip + {ENARX_SHIM_ADDRESS}]", // rsi = enclave start address mask
```
[the remark this answers is not in the document]
Comment (jarkkojs) [reply internal/shim-sgx/src/main.rs:171]: `cargo test` passed.

[the remark this answers is not in the document]
Comment (jarkkojs) [reply internal/shim-sgx/src/main.rs:171]: `HIDDED` marks it as a static symbol.

**Review by haraldh — APPROVED** [review]

**Pull request merged**